### PR TITLE
Fix PlatformTests

### PR DESCRIPTION
**Pull request by hartmark**
Fixed the failing tests in PlatformTest.

The AppDomain.CurrentDomain.BaseDirectory property was a bit of a gotcha because I got a trailing backslash from it when debugging the game, but not in the test.

Perhaps the BUG-comment is redundant, if it is missing the separator we just add it in the property.

### CODE REVIEW

File: OpenRA.Game/Platform.cs [hunk 1]
```diff
@@ -94,12 +94,25 @@ static string GetSupportDir()
...
+		{
+			get
+			{
+				var gameDir = AppDomain.CurrentDomain.BaseDirectory;
+
+				// BUG: In Game.cs we get GameDir with trailing DirectorySeparator, but not in PlatformTest.cs
+				if (!gameDir.EndsWith(Path.DirectorySeparatorChar.ToString()))
+				{
+					gameDir += Path.DirectorySeparatorChar;
+				}
```
Comment (reaperrr) [inline OpenRA.Game/Platform.cs]: > OpenRA.Game/Platform.cs:L107: [SA1513] Statements or elements wrapped in curly brackets must be followed by a blank line.

**Comment by pchote** [comment]
Can you provide logs / explain how the tests are failing?  They work fine for me (mono 5.0.1.1 on macOS 10.12)

File: OpenRA.Game/Platform.cs (hunk 1 above)
Comment (hartmark) [inline OpenRA.Game/Platform.cs]: My bad, I have forgotten to add stylecop to my devenv.

**Comment by hartmark** [comment]
Two tests fails:

  Expected string length 43 but was 42. Strings differ at index 34.
  Expected: "C:\\Users\\musku\\Source\\Repos\\OpenRA\\testpath"
  But was:  "C:\\Users\\musku\\Source\\Repos\\OpenRAtestpath"
  --------------------------------------------------^

at OpenRA.Test.PlatformTest.ResolvePath() in C:\Users\musku\Source\Repos\OpenRA\OpenRA.Test\OpenRA.Game\PlatformTest.cs:line 36



  Expected string length 10 but was 11. Strings differ at index 2.
  Expected: "./testpath"
  But was:  "./\\testpath"
  -------------^

at OpenRA.Test.PlatformTest.UnresolvePath() in C:\Users\musku\Source\Repos\OpenRA\OpenRA.Test\OpenRA.Game\PlatformTest.cs:line 52

**Comment by hartmark** [comment]
I'm running Windows 10, Visual Studio 2017

**Comment by hartmark** [comment]
@pchote any more details I need to add?

**Comment by rob-v** [comment]
@hartmark It seems your changes are not bad, but also not necessary. The tests pass successfully on my Windows 10+VS 2017.

**Comment by hartmark** [comment]
Alright, I wonder why we get different outcome on Win10.

I found another one having same problem
https://stackoverflow.com/questions/11720587/why-does-system-appdomain-currentdomain-basedirectory-return-different-results

One guy answered that it has something to do with the appdomains

After digging some more it seems that it has been a source to problems that the apppool is initialized differently.

Here's a bug for corecrl where they say it's a error in the host.
https://github.com/dotnet/coreclr/issues/10920
And that leads me to diagnose how I run the unittests.

It seems to run properly if I run it using visual studio's mstest-tool (had to add NUnit3TestAdapter nuget reference, are we not using Nuget-packages btw?)

And if I run with resharper's testrunner it fails with the error I have written above.

How are you guys running the unit-tests?
@pchote @rob-v @reaperrr

**Comment by pchote** [comment]
`make nunit`, which runs `nunit3-console`.

**Comment by hartmark** [comment]
Alright, seems the make.cmd and make.ps1 both lack the nunit command.

**Comment by aydjay** [comment]
I have also experienced this test failure when running the test using the resharper test runner. I couldn't figure our why the results were differing however.

**Comment by pchote** [comment]
I think it is reasonable to force a trailing path separator on `GameDir`.  We recently did this for `SupportDir` in #13525.  The changes to `UnresolvePath` should not be needed though, since you are guaranteed to have the separator removed by the substring.

File: OpenRA.Game/Platform.cs (hunk 1 above)
Comment (pchote) [inline OpenRA.Game/Platform.cs]: Our code style is to omit the braces when there is only a single statement.

File: OpenRA.Game/Platform.cs [hunk 2]
```diff
@@ -94,12 +94,26 @@ static string GetSupportDir()
...
 
-		public static string GameDir { get { return AppDomain.CurrentDomain.BaseDirectory; } }
+		public static string GameDir
+		{
+			get
+			{
+				var gameDir = AppDomain.CurrentDomain.BaseDirectory;
+
+				// BUG: In Game.cs we get GameDir with trailing DirectorySeparator, but not in PlatformTest.cs
+				if (!gameDir.EndsWith(Path.DirectorySeparatorChar.ToString()))
```
Comment (pchote) [inline OpenRA.Game/Platform.cs]: I suggest `if (gameDir[gameDir.Length - 1] != Path.DirectorySeparatorChar)`.  If you insist on using `EndsWith` (but IMO having to ToString the separator char is ugly) then you must include the `StringComparison.Ordinal`.

Comment (rob-v) [inline OpenRA.Game/Platform.cs]: The comment can be also corrected, as now it is incorrect - confusing as usually BaseDirectory contains trailing \ and this is workaround for a buggy? host.

File: OpenRA.Game/Platform.cs (hunk 1 above)
Comment (hartmark) [inline OpenRA.Game/Platform.cs]: Alright, I'll fix.
make check was happy however.

Enter command: check
Checking for explicit interface violations...
Checking for code style violations in OpenRA.Platforms.Default...
No violations encountered in OpenRA.Platforms.Default.
Checking for code style violations in OpenRA.Game...
No violations encountered in OpenRA.Game.
Checking for code style violations in OpenRA.Mods.Common...
No violations encountered in OpenRA.Mods.Common.
Checking for code style violations in OpenRA.Mods.Cnc...
No violations encountered in OpenRA.Mods.Cnc.
Checking for code style violations in OpenRA.Mods.D2k...
No violations encountered in OpenRA.Mods.D2k.
Checking for code style violations in OpenRA.Utility...
No violations encountered in OpenRA.Utility.
Checking for code style violations in OpenRA.Test...
No violations encountered in OpenRA.Test.
Press enter to continue.

File: OpenRA.Game/Platform.cs (hunk 2 above)
Comment (pchote) [inline OpenRA.Game/Platform.cs]: My first suggestion has the potential to behave badly with unicode paths.  Best to stick with `EndsWith` but add the `StringComparison` argument as I noted in my second suggestion.

File: OpenRA.Game/Platform.cs (hunk 1 above)
Comment (hartmark) [inline OpenRA.Game/Platform.cs]: do we have any style cop settings file? or are we just using default rules?

**Comment by hartmark** [comment]
I have updated the code per your feedback
@pchote

File: OpenRA.Game/Platform.cs (hunk 1 above)
Comment (rob-v) [inline OpenRA.Game/Platform.cs]: There is https://github.com/OpenRA/OpenRA/blob/bleed/Settings.StyleCop, but it doesn't catch all. I don't know if it isn't up to date or it doesn't support all our conventions.

Comment (hartmark) [inline OpenRA.Game/Platform.cs]: Alright, I'll check it out and add a new PR. There you can have visual studio use the file when compiling so you'll get errors for any violations.

File: .gitignore [hunk 3]
```diff
@@ -11,6 +11,7 @@ obj
 *.CodeAnalysisLog.xml
 *.lastcodeanalysissucceeded
 _ReSharper.*/
+/.vs
```
Comment (pchote) [inline .gitignore]: Can you please drop or split this to its own commit?

**Review by pchote — APPROVED** [review]
LGTM now :+1:

Just one minor nit:

**Comment by hartmark** [comment]
alright, I'll put that in a separate PR

**Comment by pchote** [comment]
Same PR is fine, just as a separate commit here.  Our policy is that a single commit shouldn't do more than one logical change.

**Comment by hartmark** [comment]
Aha, alright

File: OpenRA.Game/Platform.cs [hunk 4]
```diff
@@ -94,12 +94,24 @@ static string GetSupportDir()
 			return dir + Path.DirectorySeparatorChar;
 		}
 
-		public static string GameDir { get { return AppDomain.CurrentDomain.BaseDirectory; } }
+		public static string GameDir { get { return gameDir.Value; } }
+		static Lazy<string> gameDir = Exts.Lazy(GetGameDir);
```
Comment (rob-v) [inline OpenRA.Game/Platform.cs]: `Lazy` for GetGameDir is unnecessary. GetGameDir is called always, only once anyway, is trivial property, IMO can/should be reverted.

Comment (hartmark) [inline OpenRA.Game/Platform.cs]: Alright, I saw the other dir-properties having that so I guessed it wouldn't hurt having it on GameDir too

**Comment by hartmark** [comment]
Updated after @rob-v code comment

**Review by rob-v — APPROVED** [review]
I apologize @hartmark, I overlooked it, your change with Lazy was fine in fact. With or without it, 👍 .

**Comment by pchote** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/abd60d37867e10076ebccc938692803325fb53cb
